{"nbformat":4,"nbformat_minor":0,"metadata":{"colab":{"name":"GeneMatching.ipynb","provenance":[],"collapsed_sections":[]},"kernelspec":{"name":"python3","display_name":"Python 3"},"language_info":{"name":"python"}},"cells":[{"cell_type":"code","execution_count":null,"metadata":{"colab":{"base_uri":"https://localhost:8080/"},"id":"AOfCcj-em5_H","outputId":"afca60c2-3198-4227-c3e2-c4ff776577ea","executionInfo":{"status":"ok","timestamp":1651500475548,"user_tz":240,"elapsed":10874,"user":{"displayName":"Carolin Schulte","userId":"04929166097189432726"}}},"outputs":[{"output_type":"stream","name":"stdout","text":["Collecting biopython\n","  Downloading biopython-1.79-cp37-cp37m-manylinux_2_5_x86_64.manylinux1_x86_64.whl (2.3 MB)\n","\u001b[K     |████████████████████████████████| 2.3 MB 4.5 MB/s \n","\u001b[?25hRequirement already satisfied: numpy in /usr/local/lib/python3.7/dist-packages (from biopython) (1.21.6)\n","Installing collected packages: biopython\n","Successfully installed biopython-1.79\n"]}],"source":["!pip install biopython"]},{"cell_type":"code","source":["import Bio\n","from Bio import SeqIO, SearchIO, Entrez\n","from Bio.Seq import Seq\n","from Bio.SeqUtils import GC\n","from Bio.Blast import NCBIWWW\n","from Bio.Data import CodonTable\n","import pandas as pd\n","import numpy as np"],"metadata":{"id":"qp8heMNxnQr_"},"execution_count":null,"outputs":[]},{"cell_type":"code","source":["Entrez.email = \"cschulte@hsph.harvard.edu\""],"metadata":{"id":"dfUsOX-qnRN0"},"execution_count":null,"outputs":[]},{"cell_type":"code","source":["genes = pd.read_csv('dataset_filtered.csv')\n","info = pd.read_csv('datainfo_filtered.csv')"],"metadata":{"colab":{"base_uri":"https://localhost:8080/"},"id":"gi7DQUITnHdw","outputId":"cd4e954d-c238-46b3-9fff-e1bf2d1d0759","executionInfo":{"status":"ok","timestamp":1651501617148,"user_tz":240,"elapsed":407,"user":{"displayName":"Carolin Schulte","userId":"04929166097189432726"}}},"execution_count":null,"outputs":[{"output_type":"stream","name":"stderr","text":["/usr/local/lib/python3.7/dist-packages/IPython/core/interactiveshell.py:2882: DtypeWarning: Columns (4,5,8) have mixed types.Specify dtype option on import or set low_memory=False.\n","  exec(code_obj, self.user_global_ns, self.user_ns)\n"]}]},{"cell_type":"code","source":["info"],"metadata":{"colab":{"base_uri":"https://localhost:8080/","height":921},"id":"2049A71wnZS6","outputId":"9442807f-3fc0-400c-a352-f66c0e51ca44","executionInfo":{"status":"ok","timestamp":1651500483685,"user_tz":240,"elapsed":215,"user":{"displayName":"Carolin Schulte","userId":"04929166097189432726"}}},"execution_count":null,"outputs":[{"output_type":"execute_result","data":{"text/plain":["     Unnamed: 0  datasetID      taxID    genomeAcc  \\\n","0             1        104     862962   FQ312004.1   \n","1             2        105     226186   AE015928.1   \n","2             3        106     272560   BX571965.1   \n","3             4        107     271848   CP000086.1   \n","4             5        108      83333  NC_000913.3   \n","..          ...        ...        ...          ...   \n","107         104        221  1124478_2  NC_002506.1   \n","108         105       1221     192222  NC_002163.1   \n","109         106       1222     192222  NC_002163.1   \n","110         107       1223     192222  NC_002163.1   \n","111         108       1224     192222  NC_002163.1   \n","\n","                                       SpeciesName Classification Gram  \\\n","0                        Bacteroides fragilis 638R       Bacteria   GN   \n","1            Bacteroides thetaiotaomicron VPI-5482       Bacteria   GN   \n","2                 Burkholderia pseudomallei K96243       Bacteria   GN   \n","3                  Burkholderia thailandensis E264       Bacteria   GN   \n","4                            Escherichia coli K-12       Bacteria   GN   \n","..                                             ...            ...  ...   \n","107                  Vibrio cholerae O1 str. C6706       Bacteria   GN   \n","108  Campylobacter jejuni subsp. jejuni NCTC 11168       Bacteria   GN   \n","109  Campylobacter jejuni subsp. jejuni NCTC 11168       Bacteria   GN   \n","110  Campylobacter jejuni subsp. jejuni NCTC 11168       Bacteria   GN   \n","111  Campylobacter jejuni subsp. jejuni NCTC 11168       Bacteria   GN   \n","\n","                  Taxonomy      dataSource  \\\n","0            Bacteroidetes  PMID: 24899126   \n","1            Bacteroidetes  PMID: 19748469   \n","2       Betaproteobacteria  PMID: 24520057   \n","3       Betaproteobacteria  PMID: 23382856   \n","4      Gammaproteobacteria             PEC   \n","..                     ...             ...   \n","107    Gammaproteobacteria  PMID: 23901011   \n","108  Epsilonproteobacteria  PMID: 28806924   \n","109  Epsilonproteobacteria  PMID: 28806924   \n","110  Epsilonproteobacteria  PMID: 28806924   \n","111  Epsilonproteobacteria  PMID: 22044676   \n","\n","                                              url                dataType  \\\n","0    https://www.ncbi.nlm.nih.gov/pubmed/24899126  LargescaleAllAvailable   \n","1    https://www.ncbi.nlm.nih.gov/pubmed/19748469  LargescaleAllAvailable   \n","2    https://www.ncbi.nlm.nih.gov/pubmed/24520057  LargescaleAllAvailable   \n","3    https://www.ncbi.nlm.nih.gov/pubmed/23382856  LargescaleAllAvailable   \n","4             https://shigen.nig.ac.jp/ecoli/pec/  LargescaleAllAvailable   \n","..                                            ...                     ...   \n","107   http://www.ncbi.nlm.nih.gov/pubmed/23901011  LargescaleAllAvailable   \n","108  https://www.ncbi.nlm.nih.gov/pubmed/28806924  LargescaleAllAvailable   \n","109  https://www.ncbi.nlm.nih.gov/pubmed/28806924  LargescaleAllAvailable   \n","110  https://www.ncbi.nlm.nih.gov/pubmed/28806924  LargescaleAllAvailable   \n","111  https://www.ncbi.nlm.nih.gov/pubmed/22044676  LargescaleAllAvailable   \n","\n","            dataSubType                              experimentalTechBrief  \\\n","0                Coding                   saturated transposon mutagenesis   \n","1                Coding  insertion-sequencing, or INSeq based on a muta...   \n","2                Coding  transposon-directed insertion site sequencing ...   \n","3    CodingAndNoncoding    saturation-level transposon mutagenesis, Tn-seq   \n","4                Coding                      random transposon mutagenesis   \n","..                  ...                                                ...   \n","107  CodingAndNoncoding  hidden Markov modelÇƒÏbased analyses of transp...   \n","108              Coding                             transposon mutagenesis   \n","109              Coding                             transposon mutagenesis   \n","110              Coding                             transposon mutagenesis   \n","111              Coding  Flux Balance Analysis (FBA) and transposon mut...   \n","\n","                            experimentalConditionBrief  \\\n","0    BHI/gentamycin (25 ?g/ml)/erythromycin (10 ?g/...   \n","1    liquid TYG medium, supplemented with 10% horse...   \n","2    conjugated by Escherichia coli 19851 (pir+), s...   \n","3    conjuagted by E. coli strain SM10/?pir of suic...   \n","4                                   various conditions   \n","..                                                 ...   \n","107                    grown LB Miller (1% NaCl) media   \n","108  suplemented Muller-Hinton (MH) with Trimethopr...   \n","109  suplemented Muller-Hinton (MH) with Trimethopr...   \n","110  suplemented Muller-Hinton (MH) with Trimethopr...   \n","111  supplemented Blood Agar Base no.2 (Oxoid) with...   \n","\n","                                 definitionOfEssential  \\\n","0    genes required by B. fragilis for survival in ...   \n","1    genes required for maximal exponential growth ...   \n","2    genes required for bacterial growth (optical d...   \n","3    genes received fewer than three insertions per...   \n","4    Genes whose mutants cannot be obtained from th...   \n","..                                                 ...   \n","107                         based on insertion density   \n","108  genes required for optimal growth on MH agar p...   \n","109  genes required for optimal growth on MH agar p...   \n","110  genes required for optimal growth on MH agar p...   \n","111  A gene was considered essential if†Gr†<= 10-9†...   \n","\n","                                                  note         DateAdded  \n","0                                                  NaN  28/11/2019 00:00  \n","1                                                  NaN  28/11/2019 00:00  \n","2                                                  NaN  28/11/2019 00:00  \n","3                                                  NaN  28/11/2019 00:00  \n","4                                                  NaN  28/11/2019 00:00  \n","..                                                 ...               ...  \n","107  please note: E=essential genes, NE=nonessentia...  13/06/2016 12:10  \n","108                                                NaN  28/11/2019 00:00  \n","109                                                NaN  28/11/2019 00:00  \n","110                                                NaN  28/11/2019 00:00  \n","111                                                NaN  28/11/2019 00:00  \n","\n","[112 rows x 17 columns]"],"text/html":["\n","  <div id=\"df-d0e67bf2-94fa-4c94-aee8-90f9aa972318\">\n","    <div class=\"colab-df-container\">\n","      <div>\n","<style scoped>\n","    .dataframe tbody tr th:only-of-type {\n","        vertical-align: middle;\n","    }\n","\n","    .dataframe tbody tr th {\n","        vertical-align: top;\n","    }\n","\n","    .dataframe thead th {\n","        text-align: right;\n","    }\n","</style>\n","<table border=\"1\" class=\"dataframe\">\n","  <thead>\n","    <tr style=\"text-align: right;\">\n","      <th></th>\n","      <th>Unnamed: 0</th>\n","      <th>datasetID</th>\n","      <th>taxID</th>\n","      <th>genomeAcc</th>\n","      <th>SpeciesName</th>\n","      <th>Classification</th>\n","      <th>Gram</th>\n","      <th>Taxonomy</th>\n","      <th>dataSource</th>\n","      <th>url</th>\n","      <th>dataType</th>\n","      <th>dataSubType</th>\n","      <th>experimentalTechBrief</th>\n","      <th>experimentalConditionBrief</th>\n","      <th>definitionOfEssential</th>\n","      <th>note</th>\n","      <th>DateAdded</th>\n","    </tr>\n","  </thead>\n","  <tbody>\n","    <tr>\n","      <th>0</th>\n","      <td>1</td>\n","      <td>104</td>\n","      <td>862962</td>\n","      <td>FQ312004.1</td>\n","      <td>Bacteroides fragilis 638R</td>\n","      <td>Bacteria</td>\n","      <td>GN</td>\n","      <td>Bacteroidetes</td>\n","      <td>PMID: 24899126</td>\n","      <td>https://www.ncbi.nlm.nih.gov/pubmed/24899126</td>\n","      <td>LargescaleAllAvailable</td>\n","      <td>Coding</td>\n","      <td>saturated transposon mutagenesis</td>\n","      <td>BHI/gentamycin (25 ?g/ml)/erythromycin (10 ?g/...</td>\n","      <td>genes required by B. fragilis for survival in ...</td>\n","      <td>NaN</td>\n","      <td>28/11/2019 00:00</td>\n","    </tr>\n","    <tr>\n","      <th>1</th>\n","      <td>2</td>\n","      <td>105</td>\n","      <td>226186</td>\n","      <td>AE015928.1</td>\n","      <td>Bacteroides thetaiotaomicron VPI-5482</td>\n","      <td>Bacteria</td>\n","      <td>GN</td>\n","      <td>Bacteroidetes</td>\n","      <td>PMID: 19748469</td>\n","      <td>https://www.ncbi.nlm.nih.gov/pubmed/19748469</td>\n","      <td>LargescaleAllAvailable</td>\n","      <td>Coding</td>\n","      <td>insertion-sequencing, or INSeq based on a muta...</td>\n","      <td>liquid TYG medium, supplemented with 10% horse...</td>\n","      <td>genes required for maximal exponential growth ...</td>\n","      <td>NaN</td>\n","      <td>28/11/2019 00:00</td>\n","    </tr>\n","    <tr>\n","      <th>2</th>\n","      <td>3</td>\n","      <td>106</td>\n","      <td>272560</td>\n","      <td>BX571965.1</td>\n","      <td>Burkholderia pseudomallei K96243</td>\n","      <td>Bacteria</td>\n","      <td>GN</td>\n","      <td>Betaproteobacteria</td>\n","      <td>PMID: 24520057</td>\n","      <td>https://www.ncbi.nlm.nih.gov/pubmed/24520057</td>\n","      <td>LargescaleAllAvailable</td>\n","      <td>Coding</td>\n","      <td>transposon-directed insertion site sequencing ...</td>\n","      <td>conjugated by Escherichia coli 19851 (pir+), s...</td>\n","      <td>genes required for bacterial growth (optical d...</td>\n","      <td>NaN</td>\n","      <td>28/11/2019 00:00</td>\n","    </tr>\n","    <tr>\n","      <th>3</th>\n","      <td>4</td>\n","      <td>107</td>\n","      <td>271848</td>\n","      <td>CP000086.1</td>\n","      <td>Burkholderia thailandensis E264</td>\n","      <td>Bacteria</td>\n","      <td>GN</td>\n","      <td>Betaproteobacteria</td>\n","      <td>PMID: 23382856</td>\n","      <td>https://www.ncbi.nlm.nih.gov/pubmed/23382856</td>\n","      <td>LargescaleAllAvailable</td>\n","      <td>CodingAndNoncoding</td>\n","      <td>saturation-level transposon mutagenesis, Tn-seq</td>\n","      <td>conjuagted by E. coli strain SM10/?pir of suic...</td>\n","      <td>genes received fewer than three insertions per...</td>\n","      <td>NaN</td>\n","      <td>28/11/2019 00:00</td>\n","    </tr>\n","    <tr>\n","      <th>4</th>\n","      <td>5</td>\n","      <td>108</td>\n","      <td>83333</td>\n","      <td>NC_000913.3</td>\n","      <td>Escherichia coli K-12</td>\n","      <td>Bacteria</td>\n","      <td>GN</td>\n","      <td>Gammaproteobacteria</td>\n","      <td>PEC</td>\n","      <td>https://shigen.nig.ac.jp/ecoli/pec/</td>\n","      <td>LargescaleAllAvailable</td>\n","      <td>Coding</td>\n","      <td>random transposon mutagenesis</td>\n","      <td>various conditions</td>\n","      <td>Genes whose mutants cannot be obtained from th...</td>\n","      <td>NaN</td>\n","      <td>28/11/2019 00:00</td>\n","    </tr>\n","    <tr>\n","      <th>...</th>\n","      <td>...</td>\n","      <td>...</td>\n","      <td>...</td>\n","      <td>...</td>\n","      <td>...</td>\n","      <td>...</td>\n","      <td>...</td>\n","      <td>...</td>\n","      <td>...</td>\n","      <td>...</td>\n","      <td>...</td>\n","      <td>...</td>\n","      <td>...</td>\n","      <td>...</td>\n","      <td>...</td>\n","      <td>...</td>\n","      <td>...</td>\n","    </tr>\n","    <tr>\n","      <th>107</th>\n","      <td>104</td>\n","      <td>221</td>\n","      <td>1124478_2</td>\n","      <td>NC_002506.1</td>\n","      <td>Vibrio cholerae O1 str. C6706</td>\n","      <td>Bacteria</td>\n","      <td>GN</td>\n","      <td>Gammaproteobacteria</td>\n","      <td>PMID: 23901011</td>\n","      <td>http://www.ncbi.nlm.nih.gov/pubmed/23901011</td>\n","      <td>LargescaleAllAvailable</td>\n","      <td>CodingAndNoncoding</td>\n","      <td>hidden Markov modelÇƒÏbased analyses of transp...</td>\n","      <td>grown LB Miller (1% NaCl) media</td>\n","      <td>based on insertion density</td>\n","      <td>please note: E=essential genes, NE=nonessentia...</td>\n","      <td>13/06/2016 12:10</td>\n","    </tr>\n","    <tr>\n","      <th>108</th>\n","      <td>105</td>\n","      <td>1221</td>\n","      <td>192222</td>\n","      <td>NC_002163.1</td>\n","      <td>Campylobacter jejuni subsp. jejuni NCTC 11168</td>\n","      <td>Bacteria</td>\n","      <td>GN</td>\n","      <td>Epsilonproteobacteria</td>\n","      <td>PMID: 28806924</td>\n","      <td>https://www.ncbi.nlm.nih.gov/pubmed/28806924</td>\n","      <td>LargescaleAllAvailable</td>\n","      <td>Coding</td>\n","      <td>transposon mutagenesis</td>\n","      <td>suplemented Muller-Hinton (MH) with Trimethopr...</td>\n","      <td>genes required for optimal growth on MH agar p...</td>\n","      <td>NaN</td>\n","      <td>28/11/2019 00:00</td>\n","    </tr>\n","    <tr>\n","      <th>109</th>\n","      <td>106</td>\n","      <td>1222</td>\n","      <td>192222</td>\n","      <td>NC_002163.1</td>\n","      <td>Campylobacter jejuni subsp. jejuni NCTC 11168</td>\n","      <td>Bacteria</td>\n","      <td>GN</td>\n","      <td>Epsilonproteobacteria</td>\n","      <td>PMID: 28806924</td>\n","      <td>https://www.ncbi.nlm.nih.gov/pubmed/28806924</td>\n","      <td>LargescaleAllAvailable</td>\n","      <td>Coding</td>\n","      <td>transposon mutagenesis</td>\n","      <td>suplemented Muller-Hinton (MH) with Trimethopr...</td>\n","      <td>genes required for optimal growth on MH agar p...</td>\n","      <td>NaN</td>\n","      <td>28/11/2019 00:00</td>\n","    </tr>\n","    <tr>\n","      <th>110</th>\n","      <td>107</td>\n","      <td>1223</td>\n","      <td>192222</td>\n","      <td>NC_002163.1</td>\n","      <td>Campylobacter jejuni subsp. jejuni NCTC 11168</td>\n","      <td>Bacteria</td>\n","      <td>GN</td>\n","      <td>Epsilonproteobacteria</td>\n","      <td>PMID: 28806924</td>\n","      <td>https://www.ncbi.nlm.nih.gov/pubmed/28806924</td>\n","      <td>LargescaleAllAvailable</td>\n","      <td>Coding</td>\n","      <td>transposon mutagenesis</td>\n","      <td>suplemented Muller-Hinton (MH) with Trimethopr...</td>\n","      <td>genes required for optimal growth on MH agar p...</td>\n","      <td>NaN</td>\n","      <td>28/11/2019 00:00</td>\n","    </tr>\n","    <tr>\n","      <th>111</th>\n","      <td>108</td>\n","      <td>1224</td>\n","      <td>192222</td>\n","      <td>NC_002163.1</td>\n","      <td>Campylobacter jejuni subsp. jejuni NCTC 11168</td>\n","      <td>Bacteria</td>\n","      <td>GN</td>\n","      <td>Epsilonproteobacteria</td>\n","      <td>PMID: 22044676</td>\n","      <td>https://www.ncbi.nlm.nih.gov/pubmed/22044676</td>\n","      <td>LargescaleAllAvailable</td>\n","      <td>Coding</td>\n","      <td>Flux Balance Analysis (FBA) and transposon mut...</td>\n","      <td>supplemented Blood Agar Base no.2 (Oxoid) with...</td>\n","      <td>A gene was considered essential if†Gr†&lt;= 10-9†...</td>\n","      <td>NaN</td>\n","      <td>28/11/2019 00:00</td>\n","    </tr>\n","  </tbody>\n","</table>\n","<p>112 rows × 17 columns</p>\n","</div>\n","      <button class=\"colab-df-convert\" onclick=\"convertToInteractive('df-d0e67bf2-94fa-4c94-aee8-90f9aa972318')\"\n","              title=\"Convert this dataframe to an interactive table.\"\n","              style=\"display:none;\">\n","        \n","  <svg xmlns=\"http://www.w3.org/2000/svg\" height=\"24px\"viewBox=\"0 0 24 24\"\n","       width=\"24px\">\n","    <path d=\"M0 0h24v24H0V0z\" fill=\"none\"/>\n","    <path d=\"M18.56 5.44l.94 2.06.94-2.06 2.06-.94-2.06-.94-.94-2.06-.94 2.06-2.06.94zm-11 1L8.5 8.5l.94-2.06 2.06-.94-2.06-.94L8.5 2.5l-.94 2.06-2.06.94zm10 10l.94 2.06.94-2.06 2.06-.94-2.06-.94-.94-2.06-.94 2.06-2.06.94z\"/><path d=\"M17.41 7.96l-1.37-1.37c-.4-.4-.92-.59-1.43-.59-.52 0-1.04.2-1.43.59L10.3 9.45l-7.72 7.72c-.78.78-.78 2.05 0 2.83L4 21.41c.39.39.9.59 1.41.59.51 0 1.02-.2 1.41-.59l7.78-7.78 2.81-2.81c.8-.78.8-2.07 0-2.86zM5.41 20L4 18.59l7.72-7.72 1.47 1.35L5.41 20z\"/>\n","  </svg>\n","      </button>\n","      \n","  <style>\n","    .colab-df-container {\n","      display:flex;\n","      flex-wrap:wrap;\n","      gap: 12px;\n","    }\n","\n","    .colab-df-convert {\n","      background-color: #E8F0FE;\n","      border: none;\n","      border-radius: 50%;\n","      cursor: pointer;\n","      display: none;\n","      fill: #1967D2;\n","      height: 32px;\n","      padding: 0 0 0 0;\n","      width: 32px;\n","    }\n","\n","    .colab-df-convert:hover {\n","      background-color: #E2EBFA;\n","      box-shadow: 0px 1px 2px rgba(60, 64, 67, 0.3), 0px 1px 3px 1px rgba(60, 64, 67, 0.15);\n","      fill: #174EA6;\n","    }\n","\n","    [theme=dark] .colab-df-convert {\n","      background-color: #3B4455;\n","      fill: #D2E3FC;\n","    }\n","\n","    [theme=dark] .colab-df-convert:hover {\n","      background-color: #434B5C;\n","      box-shadow: 0px 1px 3px 1px rgba(0, 0, 0, 0.15);\n","      filter: drop-shadow(0px 1px 2px rgba(0, 0, 0, 0.3));\n","      fill: #FFFFFF;\n","    }\n","  </style>\n","\n","      <script>\n","        const buttonEl =\n","          document.querySelector('#df-d0e67bf2-94fa-4c94-aee8-90f9aa972318 button.colab-df-convert');\n","        buttonEl.style.display =\n","          google.colab.kernel.accessAllowed ? 'block' : 'none';\n","\n","        async function convertToInteractive(key) {\n","          const element = document.querySelector('#df-d0e67bf2-94fa-4c94-aee8-90f9aa972318');\n","          const dataTable =\n","            await google.colab.kernel.invokeFunction('convertToInteractive',\n","                                                     [key], {});\n","          if (!dataTable) return;\n","\n","          const docLinkHtml = 'Like what you see? Visit the ' +\n","            '<a target=\"_blank\" href=https://colab.research.google.com/notebooks/data_table.ipynb>data table notebook</a>'\n","            + ' to learn more about interactive tables.';\n","          element.innerHTML = '';\n","          dataTable['output_type'] = 'display_data';\n","          await google.colab.output.renderOutput(dataTable, element);\n","          const docLink = document.createElement('div');\n","          docLink.innerHTML = docLinkHtml;\n","          element.appendChild(docLink);\n","        }\n","      </script>\n","    </div>\n","  </div>\n","  "]},"metadata":{},"execution_count":5}]},{"cell_type":"code","source":["accList = []\n","for accession in info[\"genomeAcc\"]:\n","  try:\n","    accession = accession.split(\", \")\n","  except:\n","    continue\n","\n","  for seq in accession:\n","    if seq in accList:\n","      continue\n","    else:\n","      accList.append(seq)\n","      fetch_handle = Entrez.efetch(\"nuccore\", id=seq, rettype=\"gbwithparts\", retmode=\"text\")\n","\n","      with open(seq + \".gbk\", \"w\") as output_handle:\n","        output_handle.write(fetch_handle.read())\n","    \n","      print(\"Saved %s.gbk\" % seq)\n"],"metadata":{"colab":{"base_uri":"https://localhost:8080/"},"id":"hPlcbmcrzwRq","outputId":"e649ba74-1a6a-4570-91b6-5faa78c91e90","executionInfo":{"status":"ok","timestamp":1651504595895,"user_tz":240,"elapsed":2972203,"user":{"displayName":"Carolin Schulte","userId":"04929166097189432726"}}},"execution_count":null,"outputs":[{"output_type":"stream","name":"stdout","text":["Saved FQ312004.1.gbk\n","Saved AE015928.1.gbk\n","Saved BX571965.1.gbk\n","Saved CP000086.1.gbk\n","Saved NC_000913.3.gbk\n","Saved AL111168.1.gbk\n","Saved NC_005296.1.gbk\n","Saved NC_005297.1.gbk\n","Saved NC_007432.1.gbk\n","Saved NC_007297.2.gbk\n","Saved CP000829.1.gbk\n","Saved NC_009009.1.gbk\n","Saved LUKZ00000000.1.gbk\n","Saved NC_008577.1.gbk\n","Saved NC_008573.1.gbk\n","Saved NC_016617.1.gbk\n","Saved NC_010681.1.gbk\n","Saved NC_010676.1.gbk\n","Saved NC_010679.1.gbk\n","Saved NC_011916.1.gbk\n","Saved NC_019904.1.gbk\n","Saved NZ_CP010536.1.gbk\n","Saved NC_009952.1.gbk\n","Saved NC_009955.1.gbk\n","Saved NC_009956.1.gbk\n","Saved NC_009957.1.gbk\n","Saved NC_009958.1.gbk\n","Saved NC_009959.1.gbk\n","Saved NC_014323.1.gbk\n","Saved CP009273.1.gbk\n","Saved CABFNZ010000003.1.gbk\n","Saved NZ_CP020657.1.gbk\n","Saved CP001978.1.gbk\n","Saved NC_011769.1.gbk\n","Saved NC_004347.2.gbk\n","Saved NC_004349.1.gbk\n","Saved CP021237.1.gbk\n","Saved CP002976.1.gbk\n","Saved CP021235.1.gbk\n","Saved CP003153.1.gbk\n","Saved NZ_LUKJ01000003.1.gbk\n","Saved NC_019936.1.gbk\n","Saved NC_019937.1.gbk\n","Saved NC_019938.1.gbk\n","Saved NC_019939.1.gbk\n","Saved CP000606.1.gbk\n","Saved CP000507.1.gbk\n","Saved AL591688.1.gbk\n","Saved AL591985.1.gbk\n","Saved AE006469.1.gbk\n","Saved CP000100.1.gbk\n","Saved NZ_CP007637.1.gbk\n","Saved CR543861.1.gbk\n","Saved NC_000907.1.gbk\n","Saved NC_000962.3.gbk\n","Saved AE000511.1.gbk\n","Saved NC_007795.1.gbk\n","Saved NC_003197.2.gbk\n","Saved NC_003277.2.gbk\n","Saved AE007317.1.gbk\n","Saved NC_003062.2.gbk\n","Saved NC_003063.2.gbk\n","Saved AE009952.1.gbk\n","Saved AE014075.1.gbk\n","Saved CP000438.1.gbk\n","Saved AE004091.2.gbk\n","Saved NC_004631.1.gbk\n","Saved AE014299.2.gbk\n","Saved AM747720.1.gbk\n","Saved AM236080.1.gbk\n","Saved FQ312003.1.gbk\n","Saved AL513382.1.gbk\n","Saved NC_000964.3.gbk\n","Saved L43967.2.gbk\n","Saved NC_000912.1.gbk\n","Saved NC_002771.1.gbk\n","Saved CP000410.2.gbk\n","Saved CP000699.1.gbk\n","Saved CP000521.1.gbk\n","Saved NZ_CP009633.1.gbk\n","Saved AP009380.1.gbk\n","Saved CP003909.1.gbk\n","Saved ABQJ01000025.1.gbk\n","Saved NZ_CP043399.1.gbk\n","Saved CP002102.1.gbk\n","Saved CP001903.1.gbk\n","Saved HG328334.1.gbk\n","Saved AE003852.1.gbk\n","Saved NC_002506.1.gbk\n","Saved NC_002163.1.gbk\n"]}]},{"cell_type":"code","source":["def get_cds_feature_with_qualifier_value(seq_record, name, value):\n","    \"\"\"Function to look for CDS feature by annotation value in sequence record.\n","    \n","    e.g. You can use this for finding features by locus tag, gene ID, or protein ID.\n","    \"\"\"\n","    for feature in genome_record.features:\n","        if feature.type == \"CDS\" and value in feature.qualifiers.get(name, []):\n","            return feature\n","            \n","    return \"None\""],"metadata":{"id":"dyCk_ORhx3CX"},"execution_count":null,"outputs":[]},{"cell_type":"code","source":["cds_dict ={}\n","\n","for acc in accList:\n","  print(acc)\n","  cds_count = 0\n","  genome_record = SeqIO.read(acc + \".gbk\", 'genbank')\n","  for feature in genome_record.features:\n","    if feature.type == \"CDS\":\n","      cds_count += 1\n","  cds_dict[acc] = cds_count"],"metadata":{"colab":{"base_uri":"https://localhost:8080/"},"id":"XZiX_rjmHstw","executionInfo":{"status":"ok","timestamp":1651505136464,"user_tz":240,"elapsed":61187,"user":{"displayName":"Carolin Schulte","userId":"04929166097189432726"}},"outputId":"4fefe848-395b-4bfe-db7d-f2f510cbdead"},"execution_count":null,"outputs":[{"output_type":"stream","name":"stdout","text":["FQ312004.1\n","AE015928.1\n","BX571965.1\n","CP000086.1\n","NC_000913.3\n","AL111168.1\n","NC_005296.1\n","NC_005297.1\n","NC_007432.1\n","NC_007297.2\n","CP000829.1\n","NC_009009.1\n","LUKZ00000000.1\n","NC_008577.1\n","NC_008573.1\n","NC_016617.1\n","NC_010681.1\n","NC_010676.1\n","NC_010679.1\n","NC_011916.1\n","NC_019904.1\n","NZ_CP010536.1\n","NC_009952.1\n","NC_009955.1\n","NC_009956.1\n","NC_009957.1\n","NC_009958.1\n","NC_009959.1\n","NC_014323.1\n","CP009273.1\n","CABFNZ010000003.1\n","NZ_CP020657.1\n","CP001978.1\n","NC_011769.1\n","NC_004347.2\n","NC_004349.1\n","CP021237.1\n","CP002976.1\n","CP021235.1\n","CP003153.1\n","NZ_LUKJ01000003.1\n","NC_019936.1\n","NC_019937.1\n","NC_019938.1\n","NC_019939.1\n","CP000606.1\n","CP000507.1\n","AL591688.1\n","AL591985.1\n","AE006469.1\n","CP000100.1\n","NZ_CP007637.1\n","CR543861.1\n","NC_000907.1\n","NC_000962.3\n","AE000511.1\n","NC_007795.1\n","NC_003197.2\n","NC_003277.2\n","AE007317.1\n","NC_003062.2\n","NC_003063.2\n","AE009952.1\n","AE014075.1\n","CP000438.1\n","AE004091.2\n","NC_004631.1\n","AE014299.2\n","AM747720.1\n","AM236080.1\n","FQ312003.1\n","AL513382.1\n","NC_000964.3\n","L43967.2\n","NC_000912.1\n","NC_002771.1\n","CP000410.2\n","CP000699.1\n","CP000521.1\n","NZ_CP009633.1\n","AP009380.1\n","CP003909.1\n","ABQJ01000025.1\n","NZ_CP043399.1\n","CP002102.1\n","CP001903.1\n","HG328334.1\n","AE003852.1\n","NC_002506.1\n","NC_002163.1\n"]}]},{"cell_type":"code","source":["genes[\"gene_sequence\"] = 0\n","genes[\"protein_sequence\"] = 0\n","genes[\"strand\"] = 0\n","taxID_tmp = 0\n","\n","for i in range(len(genes)):\n","  if i%1000 == 0:\n","    print(str(round(i/len(genes)*100,2)) + \"% done\")\n","  cds_feature = \"None\"\n","  taxID = genes.at[i, \"taxaID\"]\n","  if taxID != taxID_tmp:\n","    print(taxID)\n","    taxID_tmp = taxID\n","    try:\n","      seqID = list(info[info['taxID'] == taxID][\"genomeAcc\"].str.split(r\", \"))[0][0]\n","    except:\n","      continue\n","    file = seqID + '.gbk'\n","    print(file)\n","    genome_record = SeqIO.read(file, 'genbank')\n","  \n","  gene = genes.at[i, \"gene\"]\n","  locus = genes.at[i, \"locus\"]\n","\n","  cds_feature = get_cds_feature_with_qualifier_value(genome_record, \"locus_tag\", gene)\n","  if cds_feature == \"None\":\n","    cds_feature = get_cds_feature_with_qualifier_value(genome_record, \"old_locus_tag\", gene)\n","    if cds_feature == \"None\":\n","      cds_feature = get_cds_feature_with_qualifier_value(genome_record, \"locus_tag\", locus)\n","      if cds_feature == \"None\":\n","        cds_feature = get_cds_feature_with_qualifier_value(genome_record, \"old_locus_tag\", locus)\n","  if cds_feature != \"None\":\n","    geneSeq = cds_feature.extract(genome_record.seq)\n","    try:\n","      proteinSeq = geneSeq.translate(table=11, cds=True)\n","    except:\n","      proteinSeq = 0\n","    genes.loc[i, \"gene_sequence\"] = str(geneSeq)\n","    genes.loc[i, \"protein_sequence\"] = str(proteinSeq)\n","    genes.loc[i, \"strand\"] = cds_feature.strand\n","    if i%1000 == 0:\n","      print(str(round(np.sum(genes[\"gene_sequence\"] != 0)/len(genes)*100,2)) + \"% succesfully matched\")\n"],"metadata":{"colab":{"base_uri":"https://localhost:8080/"},"id":"PV_J84DOE57P","outputId":"ab81cbf0-c1be-4200-90bf-6abb09980255","executionInfo":{"status":"ok","timestamp":1651508176546,"user_tz":240,"elapsed":2487578,"user":{"displayName":"Carolin Schulte","userId":"04929166097189432726"}}},"execution_count":null,"outputs":[{"output_type":"stream","name":"stdout","text":["0.0% done\n","862962\n","FQ312004.1.gbk\n","0.0% succesfully matched\n","226186\n","AE015928.1.gbk\n","272560\n","BX571965.1.gbk\n","0.8% done\n","0.8% succesfully matched\n","271848\n","CP000086.1.gbk\n","83333\n","NC_000913.3.gbk\n","1.6% done\n","2.4% done\n","2.15% succesfully matched\n","3.2% done\n","2.85% succesfully matched\n","4.0% done\n","3.49% succesfully matched\n","4.8% done\n","4.16% succesfully matched\n","354242\n","AL111168.1.gbk\n","5.6% done\n","4.52% succesfully matched\n","192222\n","NC_002163.1.gbk\n","6.4% done\n","5.32% succesfully matched\n","258594\n","NC_005296.1.gbk\n","7.2% done\n","6.12% succesfully matched\n","205921\n","NC_007432.1.gbk\n","8.0% done\n","6.86% succesfully matched\n","8.81% done\n","7.64% succesfully matched\n","293653\n","NC_007297.2.gbk\n","9.61% done\n","8.32% succesfully matched\n","471876\n","CP000829.1.gbk\n","10.41% done\n","8.58% succesfully matched\n","11.21% done\n","9.38% succesfully matched\n","388919\n","NC_009009.1.gbk\n","12.01% done\n","10.18% succesfully matched\n","12.81% done\n","10.95% succesfully matched\n","1813946\n","LUKZ00000000.1.gbk\n","13.61% done\n","94122\n","NC_008577.1.gbk\n","1064539\n","NC_016617.1.gbk\n","398527\n","NC_010681.1.gbk\n","14.41% done\n","11.87% succesfully matched\n","565050\n","NC_011916.1.gbk\n","926556\n","NC_019904.1.gbk\n","15.21% done\n","12.58% succesfully matched\n","68895\n","NZ_CP010536.1.gbk\n","398580\n","NC_009952.1.gbk\n","16.01% done\n","1380365\n","757424\n","NC_014323.1.gbk\n","16.81% done\n","13.5% succesfully matched\n","1122134\n","679895\n","CP009273.1.gbk\n","93064\n","CABFNZ010000003.1.gbk\n","17.61% done\n","290336\n","NZ_CP020657.1.gbk\n","225937\n","CP001978.1.gbk\n","18.41% done\n","883\n","NC_011769.1.gbk\n","211586\n","NC_004347.2.gbk\n","1986952\n","CP021237.1.gbk\n","19.21% done\n","14.25% succesfully matched\n","1423144\n","CP002976.1.gbk\n","709015\n","CP021235.1.gbk\n","20.01% done\n","14.71% succesfully matched\n","640081\n","CP003153.1.gbk\n","294\n","NZ_LUKJ01000003.1.gbk\n","20.81% done\n","21.61% done\n","644801\n","NC_019936.1.gbk\n","22.41% done\n","15.53% succesfully matched\n","323850\n","CP000606.1.gbk\n","326297\n","CP000507.1.gbk\n","266834_2\n","AL591985.1.gbk\n","266834_3\n","AE006469.1.gbk\n","266834\n","AL591688.1.gbk\n","23.21% done\n","15.84% succesfully matched\n","1140\n","CP000100.1.gbk\n","24.01% done\n","16.64% succesfully matched\n","321846\n","NZ_CP007637.1.gbk\n","1140\n","CP000100.1.gbk\n","24.82% done\n","17.05% succesfully matched\n","25.62% done\n","17.84% succesfully matched\n","26.42% done\n","62977\n","CR543861.1.gbk\n","27.22% done\n","19.36% succesfully matched\n","28.02% done\n","20.16% succesfully matched\n","71421\n","NC_000907.1.gbk\n","28.82% done\n","20.96% succesfully matched\n","29.62% done\n","21.72% succesfully matched\n","83332\n","NC_000962.3.gbk\n","30.42% done\n","22.49% succesfully matched\n","31.22% done\n","23.26% succesfully matched\n","32.02% done\n","24.04% succesfully matched\n","32.82% done\n","24.83% succesfully matched\n","83333\n","NC_000913.3.gbk\n","33.62% done\n","25.57% succesfully matched\n","34.42% done\n","26.35% succesfully matched\n","35.22% done\n","27.13% succesfully matched\n","36.02% done\n","27.9% succesfully matched\n","85962\n","AE000511.1.gbk\n","36.82% done\n","28.62% succesfully matched\n","37.62% done\n","29.42% succesfully matched\n","93061\n","NC_007795.1.gbk\n","38.42% done\n","30.17% succesfully matched\n","39.22% done\n","30.94% succesfully matched\n","40.02% done\n","31.72% succesfully matched\n","99287\n","NC_003197.2.gbk\n","171101\n","AE007317.1.gbk\n","176299\n","NC_003062.2.gbk\n","40.83% done\n","32.49% succesfully matched\n","41.63% done\n","33.23% succesfully matched\n","42.43% done\n","33.99% succesfully matched\n","176299_2\n","NC_003063.2.gbk\n","43.23% done\n","34.75% succesfully matched\n","44.03% done\n","176299\n","NC_003062.2.gbk\n","176299_2\n","NC_003063.2.gbk\n","176299\n","NC_003062.2.gbk\n","176299_2\n","NC_003063.2.gbk\n","176299\n","NC_003062.2.gbk\n","176299_2\n","NC_003063.2.gbk\n","176299\n","NC_003062.2.gbk\n","187410\n","AE009952.1.gbk\n","44.83% done\n","35.67% succesfully matched\n","45.63% done\n","36.46% succesfully matched\n","46.43% done\n","37.25% succesfully matched\n","199310\n","AE014075.1.gbk\n","47.23% done\n","37.98% succesfully matched\n","208963\n","CP000438.1.gbk\n","48.03% done\n","38.78% succesfully matched\n","48.83% done\n","39.58% succesfully matched\n","49.63% done\n","40.38% succesfully matched\n","50.43% done\n","41.18% succesfully matched\n","51.23% done\n","41.98% succesfully matched\n","52.03% done\n","42.77% succesfully matched\n","208964\n","AE004091.2.gbk\n","52.83% done\n","43.52% succesfully matched\n","53.63% done\n","44.32% succesfully matched\n","54.43% done\n","45.11% succesfully matched\n","55.23% done\n","45.91% succesfully matched\n","56.03% done\n","46.7% succesfully matched\n","209261\n","NC_004631.1.gbk\n","56.84% done\n","47.49% succesfully matched\n","57.64% done\n","48.27% succesfully matched\n","58.44% done\n","49.05% succesfully matched\n","59.24% done\n","49.82% succesfully matched\n","60.04% done\n","50.6% succesfully matched\n","211586\n","NC_004347.2.gbk\n","60.84% done\n","51.37% succesfully matched\n","61.64% done\n","52.17% succesfully matched\n","216591\n","AM747720.1.gbk\n","62.44% done\n","52.96% succesfully matched\n","63.24% done\n","53.74% succesfully matched\n","64.04% done\n","54.52% succesfully matched\n","216596\n","AM236080.1.gbk\n","64.84% done\n","55.28% succesfully matched\n","65.64% done\n","56.08% succesfully matched\n","66.44% done\n","56.88% succesfully matched\n","67.24% done\n","57.68% succesfully matched\n","68.04% done\n","58.48% succesfully matched\n","216597\n","FQ312003.1.gbk\n","68.84% done\n","59.25% succesfully matched\n","69.64% done\n","70.44% done\n","60.73% succesfully matched\n","220341\n","AL513382.1.gbk\n","71.24% done\n","72.04% done\n","72.84% done\n","73.65% done\n","74.45% done\n","224308\n","NC_000964.3.gbk\n","75.25% done\n","61.97% succesfully matched\n","76.05% done\n","62.77% succesfully matched\n","76.85% done\n","63.57% succesfully matched\n","77.65% done\n","64.36% succesfully matched\n","243273\n","L43967.2.gbk\n","272634\n","NC_000912.1.gbk\n","78.45% done\n","65.16% succesfully matched\n","79.25% done\n","272635\n","NC_002771.1.gbk\n","373153\n","CP000410.2.gbk\n","80.05% done\n","66.26% succesfully matched\n","392499\n","CP000699.1.gbk\n","400667\n","CP000521.1.gbk\n","80.85% done\n","67.02% succesfully matched\n","81.65% done\n","67.76% succesfully matched\n","82.45% done\n","68.56% succesfully matched\n","83.25% done\n","69.35% succesfully matched\n","401614\n","NZ_CP009633.1.gbk\n","84.05% done\n","431947\n","AP009380.1.gbk\n","84.85% done\n","69.57% succesfully matched\n","528354\n","CP003909.1.gbk\n","85.65% done\n","70.34% succesfully matched\n","86.45% done\n","71.11% succesfully matched\n","535206\n","ABQJ01000025.1.gbk\n","565050\n","NC_011916.1.gbk\n","87.25% done\n","71.34% succesfully matched\n","88.05% done\n","72.14% succesfully matched\n","88.85% done\n","72.94% succesfully matched\n","89.66% done\n","73.74% succesfully matched\n","588858\n","NZ_CP043399.1.gbk\n","633149\n","CP002102.1.gbk\n","90.46% done\n","73.9% succesfully matched\n","91.26% done\n","74.7% succesfully matched\n","92.06% done\n","75.5% succesfully matched\n","92.86% done\n","76.3% succesfully matched\n","714359\n","CP001903.1.gbk\n","941322\n","HG328334.1.gbk\n","93.66% done\n","94.46% done\n","95.26% done\n","96.06% done\n","96.86% done\n","1124478\n","AE003852.1.gbk\n","97.66% done\n","77.49% succesfully matched\n","98.46% done\n","78.28% succesfully matched\n","1124478_2\n","NC_002506.1.gbk\n","99.26% done\n","79.06% succesfully matched\n"]}]},{"cell_type":"code","source":["genes.shape"],"metadata":{"colab":{"base_uri":"https://localhost:8080/"},"id":"83GZrVHkJkdV","outputId":"95ec9aff-e601-4112-f7e3-4a2c29055f4a","executionInfo":{"status":"ok","timestamp":1651508243024,"user_tz":240,"elapsed":192,"user":{"displayName":"Carolin Schulte","userId":"04929166097189432726"}}},"execution_count":null,"outputs":[{"output_type":"execute_result","data":{"text/plain":["(124923, 15)"]},"metadata":{},"execution_count":38}]},{"cell_type":"code","source":["genes.to_csv('df_matched.csv')"],"metadata":{"id":"J9dgqCamNxT5"},"execution_count":null,"outputs":[]},{"cell_type":"code","source":["df = pd.read_csv('df_matched.csv')"],"metadata":{"id":"79DXRzsoL8no","colab":{"base_uri":"https://localhost:8080/"},"outputId":"57629fb8-c220-4617-9cd8-2d47f85f8440","executionInfo":{"status":"ok","timestamp":1651508257178,"user_tz":240,"elapsed":1846,"user":{"displayName":"Carolin Schulte","userId":"04929166097189432726"}}},"execution_count":null,"outputs":[{"output_type":"stream","name":"stderr","text":["/usr/local/lib/python3.7/dist-packages/IPython/core/interactiveshell.py:2882: DtypeWarning: Columns (3,5,6,9) have mixed types.Specify dtype option on import or set low_memory=False.\n","  exec(code_obj, self.user_global_ns, self.user_ns)\n"]}]},{"cell_type":"code","source":["df.head()"],"metadata":{"colab":{"base_uri":"https://localhost:8080/","height":374},"id":"m-sJLMB50og_","outputId":"d1a2478f-6840-4392-cee2-d7057aea1841","executionInfo":{"status":"ok","timestamp":1651508257179,"user_tz":240,"elapsed":16,"user":{"displayName":"Carolin Schulte","userId":"04929166097189432726"}}},"execution_count":null,"outputs":[{"output_type":"execute_result","data":{"text/plain":["   Unnamed: 0  Unnamed: 0.1  dataset  taxaID        locus gene score  \\\n","0           0             1      104  862962  BF638R_0014  NaN     1   \n","1           1             2      104  862962  BF638R_0030  NaN     1   \n","2           2             3      104  862962  BF638R_0032  NaN     1   \n","3           3             4      104  862962  BF638R_0042  NaN     1   \n","4           4             5      104  862962  BF638R_0051  NaN     1   \n","\n","  essentiality        pmid Ref_db                SpeciesName Gram  \\\n","0            E  24899126.0    NaN  Bacteroides fragilis 638R   GN   \n","1            E  24899126.0    NaN  Bacteroides fragilis 638R   GN   \n","2            E  24899126.0    NaN  Bacteroides fragilis 638R   GN   \n","3            E  24899126.0    NaN  Bacteroides fragilis 638R   GN   \n","4            E  24899126.0    NaN  Bacteroides fragilis 638R   GN   \n","\n","        Taxonomy                                      gene_sequence  \\\n","0  Bacteroidetes  ATGACTATTATTTTTCCTTCTCCTATATTCGGACCGGTTCATTCAC...   \n","1  Bacteroidetes  ATGTTACTTCTTTATATGGCATCATTATCTGAGTCCATATACCCGG...   \n","2  Bacteroidetes  ATGACAAAAGGATCTTTATTACGAAAAAGATATACATCGCAATTCT...   \n","3  Bacteroidetes  ATGAGTAAAAAGATTTTTGCGGCCCTGATAGTCGCTGTAGTCGCAA...   \n","4  Bacteroidetes  ATGAATAAACTCACTATAAATGCCTGTCCGCTATGTGGGGGCGCAC...   \n","\n","                                    protein_sequence  strand  \n","0  MTIIFPSPIFGPVHSRRLGVSLGINLLPSDGKVCSFDCIYCECGYN...       1  \n","1  MLLLYMASLSESIYPGSSARIPGLMGLCTRDDGHRYAGRDKEKTIF...      -1  \n","2  MTKGSLLRKRYTSQFCTKYYPFLLKDYTQSGCSKNHILAATTIVKI...      -1  \n","3  MSKKIFAALIVAVVATFAGYNIYQSQRVESIMSDLTMANVEALAGS...       1  \n","4  MNKLTINACPLCGGAHLKRAMTCTDFYASGEQFDLYTCEDCGFTFT...       1  "],"text/html":["\n","  <div id=\"df-77cbda7c-a7a1-4547-af74-1dfe8643d894\">\n","    <div class=\"colab-df-container\">\n","      <div>\n","<style scoped>\n","    .dataframe tbody tr th:only-of-type {\n","        vertical-align: middle;\n","    }\n","\n","    .dataframe tbody tr th {\n","        vertical-align: top;\n","    }\n","\n","    .dataframe thead th {\n","        text-align: right;\n","    }\n","</style>\n","<table border=\"1\" class=\"dataframe\">\n","  <thead>\n","    <tr style=\"text-align: right;\">\n","      <th></th>\n","      <th>Unnamed: 0</th>\n","      <th>Unnamed: 0.1</th>\n","      <th>dataset</th>\n","      <th>taxaID</th>\n","      <th>locus</th>\n","      <th>gene</th>\n","      <th>score</th>\n","      <th>essentiality</th>\n","      <th>pmid</th>\n","      <th>Ref_db</th>\n","      <th>SpeciesName</th>\n","      <th>Gram</th>\n","      <th>Taxonomy</th>\n","      <th>gene_sequence</th>\n","      <th>protein_sequence</th>\n","      <th>strand</th>\n","    </tr>\n","  </thead>\n","  <tbody>\n","    <tr>\n","      <th>0</th>\n","      <td>0</td>\n","      <td>1</td>\n","      <td>104</td>\n","      <td>862962</td>\n","      <td>BF638R_0014</td>\n","      <td>NaN</td>\n","      <td>1</td>\n","      <td>E</td>\n","      <td>24899126.0</td>\n","      <td>NaN</td>\n","      <td>Bacteroides fragilis 638R</td>\n","      <td>GN</td>\n","      <td>Bacteroidetes</td>\n","      <td>ATGACTATTATTTTTCCTTCTCCTATATTCGGACCGGTTCATTCAC...</td>\n","      <td>MTIIFPSPIFGPVHSRRLGVSLGINLLPSDGKVCSFDCIYCECGYN...</td>\n","      <td>1</td>\n","    </tr>\n","    <tr>\n","      <th>1</th>\n","      <td>1</td>\n","      <td>2</td>\n","      <td>104</td>\n","      <td>862962</td>\n","      <td>BF638R_0030</td>\n","      <td>NaN</td>\n","      <td>1</td>\n","      <td>E</td>\n","      <td>24899126.0</td>\n","      <td>NaN</td>\n","      <td>Bacteroides fragilis 638R</td>\n","      <td>GN</td>\n","      <td>Bacteroidetes</td>\n","      <td>ATGTTACTTCTTTATATGGCATCATTATCTGAGTCCATATACCCGG...</td>\n","      <td>MLLLYMASLSESIYPGSSARIPGLMGLCTRDDGHRYAGRDKEKTIF...</td>\n","      <td>-1</td>\n","    </tr>\n","    <tr>\n","      <th>2</th>\n","      <td>2</td>\n","      <td>3</td>\n","      <td>104</td>\n","      <td>862962</td>\n","      <td>BF638R_0032</td>\n","      <td>NaN</td>\n","      <td>1</td>\n","      <td>E</td>\n","      <td>24899126.0</td>\n","      <td>NaN</td>\n","      <td>Bacteroides fragilis 638R</td>\n","      <td>GN</td>\n","      <td>Bacteroidetes</td>\n","      <td>ATGACAAAAGGATCTTTATTACGAAAAAGATATACATCGCAATTCT...</td>\n","      <td>MTKGSLLRKRYTSQFCTKYYPFLLKDYTQSGCSKNHILAATTIVKI...</td>\n","      <td>-1</td>\n","    </tr>\n","    <tr>\n","      <th>3</th>\n","      <td>3</td>\n","      <td>4</td>\n","      <td>104</td>\n","      <td>862962</td>\n","      <td>BF638R_0042</td>\n","      <td>NaN</td>\n","      <td>1</td>\n","      <td>E</td>\n","      <td>24899126.0</td>\n","      <td>NaN</td>\n","      <td>Bacteroides fragilis 638R</td>\n","      <td>GN</td>\n","      <td>Bacteroidetes</td>\n","      <td>ATGAGTAAAAAGATTTTTGCGGCCCTGATAGTCGCTGTAGTCGCAA...</td>\n","      <td>MSKKIFAALIVAVVATFAGYNIYQSQRVESIMSDLTMANVEALAGS...</td>\n","      <td>1</td>\n","    </tr>\n","    <tr>\n","      <th>4</th>\n","      <td>4</td>\n","      <td>5</td>\n","      <td>104</td>\n","      <td>862962</td>\n","      <td>BF638R_0051</td>\n","      <td>NaN</td>\n","      <td>1</td>\n","      <td>E</td>\n","      <td>24899126.0</td>\n","      <td>NaN</td>\n","      <td>Bacteroides fragilis 638R</td>\n","      <td>GN</td>\n","      <td>Bacteroidetes</td>\n","      <td>ATGAATAAACTCACTATAAATGCCTGTCCGCTATGTGGGGGCGCAC...</td>\n","      <td>MNKLTINACPLCGGAHLKRAMTCTDFYASGEQFDLYTCEDCGFTFT...</td>\n","      <td>1</td>\n","    </tr>\n","  </tbody>\n","</table>\n","</div>\n","      <button class=\"colab-df-convert\" onclick=\"convertToInteractive('df-77cbda7c-a7a1-4547-af74-1dfe8643d894')\"\n","              title=\"Convert this dataframe to an interactive table.\"\n","              style=\"display:none;\">\n","        \n","  <svg xmlns=\"http://www.w3.org/2000/svg\" height=\"24px\"viewBox=\"0 0 24 24\"\n","       width=\"24px\">\n","    <path d=\"M0 0h24v24H0V0z\" fill=\"none\"/>\n","    <path d=\"M18.56 5.44l.94 2.06.94-2.06 2.06-.94-2.06-.94-.94-2.06-.94 2.06-2.06.94zm-11 1L8.5 8.5l.94-2.06 2.06-.94-2.06-.94L8.5 2.5l-.94 2.06-2.06.94zm10 10l.94 2.06.94-2.06 2.06-.94-2.06-.94-.94-2.06-.94 2.06-2.06.94z\"/><path d=\"M17.41 7.96l-1.37-1.37c-.4-.4-.92-.59-1.43-.59-.52 0-1.04.2-1.43.59L10.3 9.45l-7.72 7.72c-.78.78-.78 2.05 0 2.83L4 21.41c.39.39.9.59 1.41.59.51 0 1.02-.2 1.41-.59l7.78-7.78 2.81-2.81c.8-.78.8-2.07 0-2.86zM5.41 20L4 18.59l7.72-7.72 1.47 1.35L5.41 20z\"/>\n","  </svg>\n","      </button>\n","      \n","  <style>\n","    .colab-df-container {\n","      display:flex;\n","      flex-wrap:wrap;\n","      gap: 12px;\n","    }\n","\n","    .colab-df-convert {\n","      background-color: #E8F0FE;\n","      border: none;\n","      border-radius: 50%;\n","      cursor: pointer;\n","      display: none;\n","      fill: #1967D2;\n","      height: 32px;\n","      padding: 0 0 0 0;\n","      width: 32px;\n","    }\n","\n","    .colab-df-convert:hover {\n","      background-color: #E2EBFA;\n","      box-shadow: 0px 1px 2px rgba(60, 64, 67, 0.3), 0px 1px 3px 1px rgba(60, 64, 67, 0.15);\n","      fill: #174EA6;\n","    }\n","\n","    [theme=dark] .colab-df-convert {\n","      background-color: #3B4455;\n","      fill: #D2E3FC;\n","    }\n","\n","    [theme=dark] .colab-df-convert:hover {\n","      background-color: #434B5C;\n","      box-shadow: 0px 1px 3px 1px rgba(0, 0, 0, 0.15);\n","      filter: drop-shadow(0px 1px 2px rgba(0, 0, 0, 0.3));\n","      fill: #FFFFFF;\n","    }\n","  </style>\n","\n","      <script>\n","        const buttonEl =\n","          document.querySelector('#df-77cbda7c-a7a1-4547-af74-1dfe8643d894 button.colab-df-convert');\n","        buttonEl.style.display =\n","          google.colab.kernel.accessAllowed ? 'block' : 'none';\n","\n","        async function convertToInteractive(key) {\n","          const element = document.querySelector('#df-77cbda7c-a7a1-4547-af74-1dfe8643d894');\n","          const dataTable =\n","            await google.colab.kernel.invokeFunction('convertToInteractive',\n","                                                     [key], {});\n","          if (!dataTable) return;\n","\n","          const docLinkHtml = 'Like what you see? Visit the ' +\n","            '<a target=\"_blank\" href=https://colab.research.google.com/notebooks/data_table.ipynb>data table notebook</a>'\n","            + ' to learn more about interactive tables.';\n","          element.innerHTML = '';\n","          dataTable['output_type'] = 'display_data';\n","          await google.colab.output.renderOutput(dataTable, element);\n","          const docLink = document.createElement('div');\n","          docLink.innerHTML = docLinkHtml;\n","          element.appendChild(docLink);\n","        }\n","      </script>\n","    </div>\n","  </div>\n","  "]},"metadata":{},"execution_count":41}]},{"cell_type":"code","source":["df_final = df[df[\"gene_sequence\"] != \"0\"]"],"metadata":{"id":"MBJVjaDgxf2I"},"execution_count":null,"outputs":[]},{"cell_type":"code","source":["df_final.shape"],"metadata":{"colab":{"base_uri":"https://localhost:8080/"},"id":"as_dQIun0tBY","outputId":"aafb98f6-40e7-4aca-b1cc-d70606aa83d0","executionInfo":{"status":"ok","timestamp":1651508301321,"user_tz":240,"elapsed":141,"user":{"displayName":"Carolin Schulte","userId":"04929166097189432726"}}},"execution_count":null,"outputs":[{"output_type":"execute_result","data":{"text/plain":["(99571, 16)"]},"metadata":{},"execution_count":43}]},{"cell_type":"code","source":["len(df_final.SpeciesName.unique())"],"metadata":{"colab":{"base_uri":"https://localhost:8080/"},"id":"tVp4KAH40t8J","outputId":"e5bd7283-875b-4658-e497-79fe5b5520cb","executionInfo":{"status":"ok","timestamp":1651508306465,"user_tz":240,"elapsed":156,"user":{"displayName":"Carolin Schulte","userId":"04929166097189432726"}}},"execution_count":null,"outputs":[{"output_type":"execute_result","data":{"text/plain":["64"]},"metadata":{},"execution_count":44}]},{"cell_type":"code","source":["df_final[\"essentiality\"].value_counts()"],"metadata":{"colab":{"base_uri":"https://localhost:8080/"},"id":"WHJBFYVi1-_D","outputId":"4ba2b850-31aa-4c18-d5e3-63b73052396a","executionInfo":{"status":"ok","timestamp":1651508308942,"user_tz":240,"elapsed":154,"user":{"displayName":"Carolin Schulte","userId":"04929166097189432726"}}},"execution_count":null,"outputs":[{"output_type":"execute_result","data":{"text/plain":["NE    75178\n","E     24393\n","Name: essentiality, dtype: int64"]},"metadata":{},"execution_count":45}]},{"cell_type":"code","source":["df_final.to_csv('df_final.csv')"],"metadata":{"id":"RUUaIm0a2fQH"},"execution_count":null,"outputs":[]},{"cell_type":"code","source":["cds_dict['FQ312004.1']"],"metadata":{"colab":{"base_uri":"https://localhost:8080/"},"id":"jZDUwZLgkYcT","executionInfo":{"status":"ok","timestamp":1651513119216,"user_tz":240,"elapsed":161,"user":{"displayName":"Carolin Schulte","userId":"04929166097189432726"}},"outputId":"64bcaa49-8d0a-4a00-b3c4-bbee38f07f25"},"execution_count":null,"outputs":[{"output_type":"execute_result","data":{"text/plain":["4326"]},"metadata":{},"execution_count":50}]},{"cell_type":"code","source":[""],"metadata":{"id":"YWx5hkFX2ei4"},"execution_count":null,"outputs":[]}]}