{"nbformat":4,"nbformat_minor":0,"metadata":{"colab":{"name":"Get Orthologs.ipynb","provenance":[],"collapsed_sections":[],"authorship_tag":"ABX9TyPbXNg/dsM3mXe/9nLna8eo"},"kernelspec":{"name":"python3","display_name":"Python 3"},"language_info":{"name":"python"}},"cells":[{"cell_type":"code","execution_count":null,"metadata":{"id":"Ysm4HhR3-ziW"},"outputs":[],"source":["import os\n","import pandas as pd\n","import numpy as np"]},{"cell_type":"code","source":["from google.colab import drive\n","drive.mount('/content/drive')"],"metadata":{"colab":{"base_uri":"https://localhost:8080/"},"id":"2fsvGwv7_wsD","executionInfo":{"status":"ok","timestamp":1651685867058,"user_tz":240,"elapsed":1090,"user":{"displayName":"Ian Lo","userId":"05875249287392026449"}},"outputId":"5a916319-74c9-4de6-a467-9573fc334f25"},"execution_count":null,"outputs":[{"output_type":"stream","name":"stdout","text":["Drive already mounted at /content/drive; to attempt to forcibly remount, call drive.mount(\"/content/drive\", force_remount=True).\n"]}]},{"cell_type":"markdown","source":["# Create FASTA files of protein sequences to pass to OrthoFinder"],"metadata":{"id":"W1kO3TXQRHR5"}},{"cell_type":"code","source":["base_dir = '/content/drive/My Drive/Colab Notebooks/MIT 20.C51/Project/'\n","filename = 'df_operons.csv'\n","df = pd.read_csv(os.path.join(base_dir, filename)) \n","# filter out gene sequences whose lengths aren't multiples of 3\n","df = df[df['gene_sequence'].map(len)%3==0]\n","# filter out gene sequences that contain non ATGC letters\n","nonATGC = [any(letter not in ['A','T','G','C'] for letter in seq) for seq in df[\"gene_sequence\"]]\n","datafilter = [not elem for elem in nonATGC]\n","data = df[datafilter]\n","data.index = range(data.shape[0])"],"metadata":{"colab":{"base_uri":"https://localhost:8080/"},"id":"b45qtKQ2_CNO","executionInfo":{"status":"ok","timestamp":1651718023385,"user_tz":240,"elapsed":11953,"user":{"displayName":"Ian Lo","userId":"05875249287392026449"}},"outputId":"a3381047-f3da-4f53-d9ef-a360c5fde898"},"execution_count":98,"outputs":[{"output_type":"stream","name":"stderr","text":["/usr/local/lib/python3.7/dist-packages/IPython/core/interactiveshell.py:2882: DtypeWarning: Columns (3) have mixed types.Specify dtype option on import or set low_memory=False.\n","  exec(code_obj, self.user_global_ns, self.user_ns)\n"]}]},{"cell_type":"code","source":["data.shape"],"metadata":{"colab":{"base_uri":"https://localhost:8080/"},"id":"AtUbbmPp_zXt","executionInfo":{"status":"ok","timestamp":1651687769691,"user_tz":240,"elapsed":173,"user":{"displayName":"Ian Lo","userId":"05875249287392026449"}},"outputId":"72446ea8-e800-4515-b81a-8289374cffc6"},"execution_count":null,"outputs":[{"output_type":"execute_result","data":{"text/plain":["(92624, 13)"]},"metadata":{},"execution_count":80}]},{"cell_type":"code","source":["def writeFasta (data, filename):\n","  fastaLines = []\n","  for i in range(data.shape[0]):\n","    fastaLines.append(\">\" + data[\"locus\"][i] + \"| [\" + data[\"SpeciesName\"][i]+ \"]\")\n","    fastaLines.append(\"\\n\" + data[\"protein_sequence\"][i] + \"\\n\")\n","  # output to file\n","  with open(filename, 'w+') as f:\n","    f.writelines(fastaLines)"],"metadata":{"id":"HAdw0-4mAsZ4"},"execution_count":null,"outputs":[]},{"cell_type":"code","source":["species = data.SpeciesName.unique() #64 species\n","species[0:5]"],"metadata":{"colab":{"base_uri":"https://localhost:8080/"},"id":"9hradrrMA8zr","executionInfo":{"status":"ok","timestamp":1651686594148,"user_tz":240,"elapsed":164,"user":{"displayName":"Ian Lo","userId":"05875249287392026449"}},"outputId":"7afae83d-7bd3-4316-eb0a-e6c8132203ac"},"execution_count":null,"outputs":[{"output_type":"execute_result","data":{"text/plain":["array(['Bacteroides fragilis 638R',\n","       'Bacteroides thetaiotaomicron VPI-5482',\n","       'Burkholderia pseudomallei K96243',\n","       'Burkholderia thailandensis E264', 'Escherichia coli K-12'],\n","      dtype=object)"]},"metadata":{},"execution_count":61}]},{"cell_type":"code","source":["fasta_dir = os.path.join(base_dir, 'Fasta')\n","fasta_dir"],"metadata":{"colab":{"base_uri":"https://localhost:8080/","height":35},"id":"RIEQSyZSJNSJ","executionInfo":{"status":"ok","timestamp":1651686604080,"user_tz":240,"elapsed":6,"user":{"displayName":"Ian Lo","userId":"05875249287392026449"}},"outputId":"9c0d49e2-3539-4ff5-c3ce-4d21253fc8e9"},"execution_count":null,"outputs":[{"output_type":"execute_result","data":{"text/plain":["'/content/drive/My Drive/Colab Notebooks/MIT 20.C51/Project/Fasta'"],"application/vnd.google.colaboratory.intrinsic+json":{"type":"string"}},"metadata":{},"execution_count":63}]},{"cell_type":"code","source":["for bacteria in species:\n","  dataset = data[data[\"SpeciesName\"]==bacteria]\n","  dataset.index = range(dataset.shape[0])\n","  bac = bacteria.replace(\" \", \"_\")\n","  filename = os.path.join(fasta_dir, bac + \".faa\")\n","  writeFasta(dataset, filename)"],"metadata":{"id":"pQEn9NCYAwHc"},"execution_count":null,"outputs":[]},{"cell_type":"markdown","source":["# Process OrthoFinder results"],"metadata":{"id":"StnYmgRRROAN"}},{"cell_type":"code","source":["filename = 'Orthogroups_clean.txt'\n","ortholog = pd.read_csv(os.path.join(base_dir, filename), sep='\\t') "],"metadata":{"id":"QVe4Q6IeSYTb","executionInfo":{"status":"ok","timestamp":1651718049987,"user_tz":240,"elapsed":155,"user":{"displayName":"Ian Lo","userId":"05875249287392026449"}}},"execution_count":99,"outputs":[]},{"cell_type":"code","source":["ortholog.head()"],"metadata":{"colab":{"base_uri":"https://localhost:8080/","height":206},"id":"lbC6VXu2EWFg","executionInfo":{"status":"ok","timestamp":1651718059831,"user_tz":240,"elapsed":140,"user":{"displayName":"Ian Lo","userId":"05875249287392026449"}},"outputId":"eff3f7ff-be7b-421c-d741-ddacfcd66d99"},"execution_count":100,"outputs":[{"output_type":"execute_result","data":{"text/plain":["  Orthogroup                         SpeciesName  \\\n","0  OG0000000  Acinetobacter baumannii ATCC 17978   \n","1  OG0000000           Acinetobacter baylyi ADP1   \n","2  OG0000000       Agrobacterium fabrum str. C58   \n","3  OG0000000                  Azospira oryzae PS   \n","4  OG0000000            Azospirillum baldaniorum   \n","\n","                                            GeneLoci  \n","0  A1S_0026|A1S_0144|A1S_0986|A1S_1059|A1S_1060|A...  \n","1  ACIAD0034|ACIAD0175|ACIAD0969|ACIAD1058|ACIAD1...  \n","2  Atu0159|Atu0174|Atu0190|Atu0197|Atu0308|Atu033...  \n","3  Dsui_0467|Dsui_0705|Dsui_1167|Dsui_2857|Dsui_3...  \n","4         AZOBR_RS02260|AZOBR_RS07285|AZOBR_RS12725|  "],"text/html":["\n","  <div id=\"df-13624e95-6076-4a1f-8d94-7e0ca7171cd9\">\n","    <div class=\"colab-df-container\">\n","      <div>\n","<style scoped>\n","    .dataframe tbody tr th:only-of-type {\n","        vertical-align: middle;\n","    }\n","\n","    .dataframe tbody tr th {\n","        vertical-align: top;\n","    }\n","\n","    .dataframe thead th {\n","        text-align: right;\n","    }\n","</style>\n","<table border=\"1\" class=\"dataframe\">\n","  <thead>\n","    <tr style=\"text-align: right;\">\n","      <th></th>\n","      <th>Orthogroup</th>\n","      <th>SpeciesName</th>\n","      <th>GeneLoci</th>\n","    </tr>\n","  </thead>\n","  <tbody>\n","    <tr>\n","      <th>0</th>\n","      <td>OG0000000</td>\n","      <td>Acinetobacter baumannii ATCC 17978</td>\n","      <td>A1S_0026|A1S_0144|A1S_0986|A1S_1059|A1S_1060|A...</td>\n","    </tr>\n","    <tr>\n","      <th>1</th>\n","      <td>OG0000000</td>\n","      <td>Acinetobacter baylyi ADP1</td>\n","      <td>ACIAD0034|ACIAD0175|ACIAD0969|ACIAD1058|ACIAD1...</td>\n","    </tr>\n","    <tr>\n","      <th>2</th>\n","      <td>OG0000000</td>\n","      <td>Agrobacterium fabrum str. C58</td>\n","      <td>Atu0159|Atu0174|Atu0190|Atu0197|Atu0308|Atu033...</td>\n","    </tr>\n","    <tr>\n","      <th>3</th>\n","      <td>OG0000000</td>\n","      <td>Azospira oryzae PS</td>\n","      <td>Dsui_0467|Dsui_0705|Dsui_1167|Dsui_2857|Dsui_3...</td>\n","    </tr>\n","    <tr>\n","      <th>4</th>\n","      <td>OG0000000</td>\n","      <td>Azospirillum baldaniorum</td>\n","      <td>AZOBR_RS02260|AZOBR_RS07285|AZOBR_RS12725|</td>\n","    </tr>\n","  </tbody>\n","</table>\n","</div>\n","      <button class=\"colab-df-convert\" onclick=\"convertToInteractive('df-13624e95-6076-4a1f-8d94-7e0ca7171cd9')\"\n","              title=\"Convert this dataframe to an interactive table.\"\n","              style=\"display:none;\">\n","        \n","  <svg xmlns=\"http://www.w3.org/2000/svg\" height=\"24px\"viewBox=\"0 0 24 24\"\n","       width=\"24px\">\n","    <path d=\"M0 0h24v24H0V0z\" fill=\"none\"/>\n","    <path d=\"M18.56 5.44l.94 2.06.94-2.06 2.06-.94-2.06-.94-.94-2.06-.94 2.06-2.06.94zm-11 1L8.5 8.5l.94-2.06 2.06-.94-2.06-.94L8.5 2.5l-.94 2.06-2.06.94zm10 10l.94 2.06.94-2.06 2.06-.94-2.06-.94-.94-2.06-.94 2.06-2.06.94z\"/><path d=\"M17.41 7.96l-1.37-1.37c-.4-.4-.92-.59-1.43-.59-.52 0-1.04.2-1.43.59L10.3 9.45l-7.72 7.72c-.78.78-.78 2.05 0 2.83L4 21.41c.39.39.9.59 1.41.59.51 0 1.02-.2 1.41-.59l7.78-7.78 2.81-2.81c.8-.78.8-2.07 0-2.86zM5.41 20L4 18.59l7.72-7.72 1.47 1.35L5.41 20z\"/>\n","  </svg>\n","      </button>\n","      \n","  <style>\n","    .colab-df-container {\n","      display:flex;\n","      flex-wrap:wrap;\n","      gap: 12px;\n","    }\n","\n","    .colab-df-convert {\n","      background-color: #E8F0FE;\n","      border: none;\n","      border-radius: 50%;\n","      cursor: pointer;\n","      display: none;\n","      fill: #1967D2;\n","      height: 32px;\n","      padding: 0 0 0 0;\n","      width: 32px;\n","    }\n","\n","    .colab-df-convert:hover {\n","      background-color: #E2EBFA;\n","      box-shadow: 0px 1px 2px rgba(60, 64, 67, 0.3), 0px 1px 3px 1px rgba(60, 64, 67, 0.15);\n","      fill: #174EA6;\n","    }\n","\n","    [theme=dark] .colab-df-convert {\n","      background-color: #3B4455;\n","      fill: #D2E3FC;\n","    }\n","\n","    [theme=dark] .colab-df-convert:hover {\n","      background-color: #434B5C;\n","      box-shadow: 0px 1px 3px 1px rgba(0, 0, 0, 0.15);\n","      filter: drop-shadow(0px 1px 2px rgba(0, 0, 0, 0.3));\n","      fill: #FFFFFF;\n","    }\n","  </style>\n","\n","      <script>\n","        const buttonEl =\n","          document.querySelector('#df-13624e95-6076-4a1f-8d94-7e0ca7171cd9 button.colab-df-convert');\n","        buttonEl.style.display =\n","          google.colab.kernel.accessAllowed ? 'block' : 'none';\n","\n","        async function convertToInteractive(key) {\n","          const element = document.querySelector('#df-13624e95-6076-4a1f-8d94-7e0ca7171cd9');\n","          const dataTable =\n","            await google.colab.kernel.invokeFunction('convertToInteractive',\n","                                                     [key], {});\n","          if (!dataTable) return;\n","\n","          const docLinkHtml = 'Like what you see? Visit the ' +\n","            '<a target=\"_blank\" href=https://colab.research.google.com/notebooks/data_table.ipynb>data table notebook</a>'\n","            + ' to learn more about interactive tables.';\n","          element.innerHTML = '';\n","          dataTable['output_type'] = 'display_data';\n","          await google.colab.output.renderOutput(dataTable, element);\n","          const docLink = document.createElement('div');\n","          docLink.innerHTML = docLinkHtml;\n","          element.appendChild(docLink);\n","        }\n","      </script>\n","    </div>\n","  </div>\n","  "]},"metadata":{},"execution_count":100}]},{"cell_type":"code","source":["Orthogroup = ['temp']\n","SpeciesName = ['temp']\n","Locus = ['temp']\n","\n","for i in range(ortholog.shape[0]):\n","  loci = ortholog['GeneLoci'][i].split('|')\n","  loci.remove('')\n","  Locus.extend(loci)\n","  Orthogroup.extend([ortholog['Orthogroup'][i]] * len(loci))\n","  SpeciesName.extend([ortholog['SpeciesName'][i]] * len(loci))\n","\n","ortholog_df = pd.DataFrame(list(zip(Orthogroup, SpeciesName, Locus)),\n","                           columns =['orthogroup', 'SpeciesName','locus'])  \n","ortholog_df = ortholog_df.iloc[1:,] #remove first row"],"metadata":{"id":"AaOkn8vGEYPY","executionInfo":{"status":"ok","timestamp":1651722072336,"user_tz":240,"elapsed":1375,"user":{"displayName":"Ian Lo","userId":"05875249287392026449"}}},"execution_count":182,"outputs":[]},{"cell_type":"code","source":["ortholog_df.shape #82706 genes, matches the Statistics file"],"metadata":{"colab":{"base_uri":"https://localhost:8080/"},"id":"LgeticDVI1Ox","executionInfo":{"status":"ok","timestamp":1651722125894,"user_tz":240,"elapsed":155,"user":{"displayName":"Ian Lo","userId":"05875249287392026449"}},"outputId":"95175be4-fd2b-438c-a09f-980ca30f38d7"},"execution_count":185,"outputs":[{"output_type":"execute_result","data":{"text/plain":["(82706, 3)"]},"metadata":{},"execution_count":185}]},{"cell_type":"code","source":["# 2 SpeciesNames missing punction in ortholog file\n","ortholog_df.loc[ ortholog_df[\"SpeciesName\"] == 'Desulfovibrio vulgaris str.  Miyazaki F ', \"SpeciesName\"] = \"Desulfovibrio vulgaris str. 'Miyazaki F'\"\n","ortholog_df.loc[ ortholog_df[\"SpeciesName\"] == 'Escherichia coli O25b H4-ST131', \"SpeciesName\"] = \"Escherichia coli O25b:H4-ST131\""],"metadata":{"id":"0lZXTJxGQlij","executionInfo":{"status":"ok","timestamp":1651722121967,"user_tz":240,"elapsed":175,"user":{"displayName":"Ian Lo","userId":"05875249287392026449"}}},"execution_count":184,"outputs":[]},{"cell_type":"code","source":["# map orthogroup to original dataframe\n","df_ortholog = pd.merge(\n","              data,\n","              ortholog_df,\n","              how=\"left\",\n","              on=[\"SpeciesName\",\"locus\"]           \n","          )"],"metadata":{"id":"NwPUwvbQRfdO","executionInfo":{"status":"ok","timestamp":1651722161552,"user_tz":240,"elapsed":538,"user":{"displayName":"Ian Lo","userId":"05875249287392026449"}}},"execution_count":189,"outputs":[]},{"cell_type":"code","source":["df_ortholog.shape"],"metadata":{"colab":{"base_uri":"https://localhost:8080/"},"id":"tKI53L4oLYw4","executionInfo":{"status":"ok","timestamp":1651722166007,"user_tz":240,"elapsed":164,"user":{"displayName":"Ian Lo","userId":"05875249287392026449"}},"outputId":"9b1725ad-670a-4ae1-e611-3e0dab02b435"},"execution_count":190,"outputs":[{"output_type":"execute_result","data":{"text/plain":["(92624, 14)"]},"metadata":{},"execution_count":190}]},{"cell_type":"code","source":["# check that there are the same number of matched genes to ortholog groups in merged dataset\n","df_ortholog[df_ortholog['orthogroup'].notnull()].shape[0]"],"metadata":{"colab":{"base_uri":"https://localhost:8080/"},"id":"6z89bpoSOYZV","executionInfo":{"status":"ok","timestamp":1651722345864,"user_tz":240,"elapsed":176,"user":{"displayName":"Ian Lo","userId":"05875249287392026449"}},"outputId":"75eafb5c-95e6-4bf3-b63a-0d6286ea6085"},"execution_count":193,"outputs":[{"output_type":"execute_result","data":{"text/plain":["82706"]},"metadata":{},"execution_count":193}]},{"cell_type":"code","source":["filename = 'df_ortholog.csv'\n","df_ortholog.to_csv(os.path.join(base_dir, filename))"],"metadata":{"id":"OXNzVmNBUzo0","executionInfo":{"status":"ok","timestamp":1651722647637,"user_tz":240,"elapsed":4508,"user":{"displayName":"Ian Lo","userId":"05875249287392026449"}}},"execution_count":195,"outputs":[]}]}